storage: TestRaftRemoveRace is flaky under stress (ambiguous result)

Somewhat hard to reproduce locally, but reproduces almost instantly in GCE:
```
tamird@gceworker173:~/go/src/github.com/cockroachdb/cockroach$ make stress STRESSFLAGS='-p 100 -stderr -maxfails 1' PKG=./pkg/storage TESTS=TestRaftRemoveRace
go list -tags '' -f 'go test -v  -tags '\'''\'' -ldflags '\'''\'' -i -c {{.ImportPath}} -o {{.Dir}}/stress.test && (cd {{.Dir}} && if [ -f stress.test ]; then stress -p 100 -stderr -maxfails 1 ./stress.test -test.run '\''TestRaftRemoveRace'\'' -test.timeout 2m ; fi)' ./pkg/storage | /bin/bash
github.com/cockroachdb/cockroach/pkg/storage
github.com/cockroachdb/cockroach/pkg/storage_test
testmain
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
0 runs so far, 0 failures, over 15s
2 runs so far, 0 failures, over 20s
9 runs so far, 0 failures, over 25s
57 runs so far, 0 failures, over 30s
96 runs so far, 0 failures, over 35s
98 runs so far, 0 failures, over 40s

--- FAIL: TestRaftRemoveRace (7.48s)
	client_test.go:915: change replicas of range 1 failed: result is ambiguous (Raft re-proposal failed: write at timestamp 0.000000123,1564 too old; wrote at 0.000000123,1565)
FAIL


ERROR: exit status 1

99 runs completed, 1 failures, over 41s
FAIL
Makefile:138: recipe for target 'stress' failed
make: *** [stress] Error 1
```

My guess is that we need better error handling in either `(*storage.Replica).ChangeReplicas` or its callers `(*storage.multiTestContext).{,un}replicateRange`.

--- a/pkg/storage/client_test.go
+++ b/pkg/storage/client_test.go
@@ -54,6 +54,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/storage/engine"
 	"github.com/cockroachdb/cockroach/pkg/storage/engine/enginepb"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -903,15 +904,23 @@ func (m *multiTestContext) replicateRange(rangeID roachpb.RangeID, dests ...int)
 			m.t.Fatal(err)
 		}
 
-		if err := rep.ChangeReplicas(
-			rep.AnnotateCtx(ctx),
-			roachpb.ADD_REPLICA,
-			roachpb.ReplicaDescriptor{
-				NodeID:  m.stores[dest].Ident.NodeID,
-				StoreID: m.stores[dest].Ident.StoreID,
-			},
-			&desc,
-		); err != nil {
+		// Assume AmbiguousResultErrors are due to re-proposals and the
+		// underlying change replicas succeeded.
+		for {
+			err := rep.ChangeReplicas(
+				rep.AnnotateCtx(ctx),
+				roachpb.ADD_REPLICA,
+				roachpb.ReplicaDescriptor{
+					NodeID:  m.stores[dest].Ident.NodeID,
+					StoreID: m.stores[dest].Ident.StoreID,
+				},
+				&desc,
+			)
+			if err == nil || testutils.IsError(err, "unable to add replica .* which is already present") {
+				break
+			} else if _, ok := errors.Cause(err).(*roachpb.AmbiguousResultError); ok {
+				break
+			}
 			m.t.Fatal(err)
 		}
 
@@ -959,16 +968,24 @@ func (m *multiTestContext) unreplicateRange(rangeID roachpb.RangeID, dest int) {
 
 	ctx := rep.AnnotateCtx(context.Background())
 
-	if err := rep.ChangeReplicas(
-		ctx,
-		roachpb.REMOVE_REPLICA,
-		roachpb.ReplicaDescriptor{
-			NodeID:  m.idents[dest].NodeID,
-			StoreID: m.idents[dest].StoreID,
-		},
-		&desc,
-	); err != nil {
-		m.t.Fatal(err)
+	// Assume AmbiguousResultErrors are due to re-proposals and the
+	// underlying change replicas succeeded.
+	for {
+		err := rep.ChangeReplicas(
+			ctx,
+			roachpb.REMOVE_REPLICA,
+			roachpb.ReplicaDescriptor{
+				NodeID:  m.idents[dest].NodeID,
+				StoreID: m.idents[dest].StoreID,
+			},
+			&desc,
+		)
+		if err == nil || testutils.IsError(err, "unable to remove replica .* which is not present") {
+			break
+		} else if _, ok := errors.Cause(err).(*roachpb.AmbiguousResultError); ok {
+			break
+		}
+		m.t.Fatalf("failed: %T, %s", err, err)
 	}
 }
 

